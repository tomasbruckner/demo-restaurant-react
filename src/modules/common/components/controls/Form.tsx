import React, { PureComponent } from 'react';
import styled from 'styled-components/macro';
import { CircularProgress } from '@material-ui/core';

const StyledForm = styled.form<{ fullWidth?: boolean }>`
  max-width: ${p => (p.fullWidth ? 'auto' : '500px')};
`;

const LoadingOverlay = styled.div`
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  justify-content: center;
  position: absolute;
  z-index: 100;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.2);
`;

export default class Form extends PureComponent<{
  onSubmit?: Function;
  onSubmitError?: Function;
  fullWidth?: boolean;
}> {
  state = {
    isLoading: false,
  };

  onSubmit = async (ev: any) => {
    ev.preventDefault();
    const { onSubmit, onSubmitError } = this.props;

    if (!onSubmit) {
      return;
    }

    this.setState({ isLoading: true });
    try {
      await onSubmit(ev);
    } catch (e) {
      if (onSubmitError) {
        onSubmitError();
      }
    }

    this.setState({ isLoading: false });
  };

  render() {
    const { children, fullWidth } = this.props;
    const { isLoading } = this.state;

    return (
      <StyledForm id="styledFormId" onSubmit={this.onSubmit} fullWidth={fullWidth}>
        {isLoading && (
          <LoadingOverlay>
            <CircularProgress color="primary" />
          </LoadingOverlay>
        )}
        {children}
      </StyledForm>
    );
  }
}
