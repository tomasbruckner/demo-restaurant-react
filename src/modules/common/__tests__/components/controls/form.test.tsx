import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import Form from 'modules/common/components/controls/Form';

describe('Form', () => {
  it('basic example', () => {
    const tree = renderer.create(<Form />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
