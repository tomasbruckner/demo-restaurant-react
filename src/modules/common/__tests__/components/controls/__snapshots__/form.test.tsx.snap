// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form basic example 1`] = `
.c0 {
  max-width: 500px;
}

<form
  className="c0"
  id="styledFormId"
  onSubmit={[Function]}
/>
`;
